### PR TITLE
Add the sha3 hash method support in werkzeug.security

### PR DESCRIPTION
**Pull request by wwqgtxx**

### CODE REVIEW

File: werkzeug/security.py [hunk 1]
```diff
@@ -34,7 +34,11 @@
 
 
 def _find_hashlib_algorithms():
-    algos = getattr(hashlib, 'algorithms', None)
+    #for python2>=2.7.9 and python3>=3.2
```
Comment (ThiefMaster) [inline werkzeug/security.py]: PEP8: space after `#`

File: werkzeug/security.py [hunk 2]
```diff
@@ -34,7 +34,11 @@
...
+    algos = getattr(hashlib, 'algorithms_available', None)
+    #for 2.7.0 < python2 < 2.7.9
```
Comment (ThiefMaster) [inline werkzeug/security.py]: same.

File: werkzeug/security.py [hunk 3]
```diff
@@ -99,9 +103,14 @@ def pbkdf2_bin(data, salt, iterations=DEFAULT_PBKDF2_ITERATIONS,
...
         if hasattr(_test_hash, 'name') and \
            _test_hash.name in _hash_funcs:
-            return hashlib.pbkdf2_hmac(_test_hash.name,
+            try:
+                return hashlib.pbkdf2_hmac(_test_hash.name,
                                        data, salt, iterations,
                                        keylen)
+            # python3's hashlib.pbkdf2_hmac OpenSSL version not support the sha3 hash method
+            # need use the pure python implementation
+            except ValueError:
```
Comment (ThiefMaster) [inline werkzeug/security.py]: this feels too generic for me, especially in security-related code.

File: werkzeug/security.py [hunk 4]
```diff
@@ -99,9 +103,14 @@ def pbkdf2_bin(data, salt, iterations=DEFAULT_PBKDF2_ITERATIONS,
         _test_hash = hashfunc()
         if hasattr(_test_hash, 'name') and \
            _test_hash.name in _hash_funcs:
-            return hashlib.pbkdf2_hmac(_test_hash.name,
+            try:
+                return hashlib.pbkdf2_hmac(_test_hash.name,
                                        data, salt, iterations,
                                        keylen)
+            # python3's hashlib.pbkdf2_hmac OpenSSL version not support the sha3 hash method
```
Comment (ThiefMaster) [inline werkzeug/security.py]: *does* not support

File: werkzeug/security.py [hunk 5]
```diff
@@ -99,9 +103,14 @@ def pbkdf2_bin(data, salt, iterations=DEFAULT_PBKDF2_ITERATIONS,
...
+            # need use the pure python implementation
```
Comment (ThiefMaster) [inline werkzeug/security.py]: *so we need to use the pure python implementation*

**Comment by lepture** [comment]
This could be fixed by https://github.com/pallets/werkzeug/pull/1282

**Pull request closed without merging**